### ISSUE TITLE
Looking for a better way to do things on behalf of another subscription

### ISSUE THREAD

**Issue by cristian-medina** [Azure/azure-sdk-for-python#838]
If I have multiple Azure accounts and multiple subscriptions, and I'm trying to manage all of these with the same backend application using this SDK, I need to go into each subscription and perform the following steps to retrieve the highlighted information:
- Go to the Azure Portal and create a new App Registration.
- When that is complete look at the details of the application and get the **application id**.
- Go to the _Keys_ section, generate a new **secret key** and store that number.
- Go to the _Endpoints_ section - this is very not intuitive btw - and examine the URLs that come up to retrieve the part of the URL that acts as my **tenant id**.
- Look up my subscription details and pull the **subscription id**.

Once all of this is complete, execute an Azure CLI command or PowerShell script that will then grant permissions needed for the new application to do its job.

These are A LOT of numbers to store and keep track of for each subscription. I'd like to see something closer to what Amazon AWS does with IAM Roles, where each account that wants to grant you permissions creates an IAM _Role_ and gives the account that's doing the work a global identifier for that role. That process that does the work then authenticates to Amazon with its own set of secret keys, and performs an _assumeRole_ action to take on the credentials granted by the identifier for that role.

Beyond just permissions, the identifier itself represents a link between two accounts, and the assumeRole action also takes an "external id" field which can be used to further validate the account that's trying to assume those permissions.

This would greatly reduce the complexity of monitoring and management software that we may write for our organizations.

**Comment by lmazuel** [comment]
Hi @cristian-medina thanks for your feedback!

@yugangw-msft what do you think here?

**Comment by robotechredmond** [comment]
@cristian-medina Thanks for your feedback! We'll pass this along to the Azure engineering team that managed RBAC for review.  In the meantime, there's a similar pattern that can be built for performing something similar to IAM roles using Azure AD + Azure Key Vault.  Here's a link to more details: http://www.dushyantgill.com/blog/2015/04/26/say-goodbye-to-key-management-manage-access-to-azure-storage-data-using-azure-ad/

**Comment by yugangw-msft** [comment]
@cristian-medina, not close to the iam roles you prefer, but hope it gets your existing work flow easier with azure.  I suggets you checkout [azure-cli](https://github.com/Azure/azure-cli), particularly the following command
1. `az ad sp create-for-rbac`: will do all your portal configuration work, plus you can create any number of assignments at the same time
2. In the help of 'create-for-rbac', it points out a `login` command, which you can run to login.
3. Run `az account list --query "[].id" -o tsv`, which should give you all the subsccription

**Comment by cristian-medina** [comment]
@yugangw-msft this command does reduce the number of steps required, thanks. 

Took me a bit to realize that whatever I pass as `--name` maps to the `client_id` when using `ServicePrincipalCredentials` in python to login, and that I can use the `password` from the response as the `secret`. This was mostly because the parameters are different from what you get if creating things through the UI.

**Comment by yugangw-msft** [comment]
@cristian-medina, we might have terminology inconsistency between cli and sdk. `--name` gets mapped to `APP ID URI` you saw in portal. This value and `CLIENT ID` (essentially the `APP ID`)  both are in the the array of `servicePrincipalNames` in the core graph project, and you can use either one, along with `secret` (named as `password` in cli) to acquire a token.
Hope this clarifies.

**Comment by lmazuel** [comment]
I close this issue since I truly think we answered the initial question. Please feel free to comment on it if you disagree and I can re-open (please descibe your current status in that situation) or open a new issue if necessary.
Thank you!
